CI: create Docker container containing compiled tools

aparcar opened the pull request:
Currently each Kernel compilation takes about 30 minutes of which 20
minutes are used to copile our tools. While the toolchain is downloaded
and instantly ready the tools are missing.

This commit starts uploading a Docker container including compiled tools
which are ready to use. It is automatically updated whenever any tools
are changed.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
CI: create Docker container containing compiled tools

Currently each Kernel compilation takes about 30 minutes of which 20
minutes are used to compile our tools. While the toolchain is downloaded
and instantly ready the tools are missing.

This commit starts uploading a Docker container including compiled tools
which are ready to use. It is automatically updated whenever any tools
are changed.

Signed-off-by: Paul Spooren <mail@aparcar.org>
Co-Developed-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
Oh nice!

Ansuel commented:
Anyway I'm checking the tools-ubuntu-latest action and I notice the error.
What was the idea? To put the openwrt directory in the container? 

Wonder if a better solution would be tar the tools present in the staging dir and extract them from the container?

AFAIK github actions have some limitation so the openwrt dir is placed in a strange path and that can't be changed so shipping the openwrt dir with the container seems problematic to me.

Ansuel commented:
@aparcar news?

aparcar commented:
I tried to create the container but it seems that it always rebuilds the tools, rendering such container obsolete. I'm trying to figure out what's the issue here

Ansuel commented:
@aparcar mhhh where are you putting the tools? are you sure they are actually used by the openwrt repo that is cloned by the github action?

Ansuel commented:
(testing some thing and didn't notice i was pushing to this branch... sorry for the mess reverted...)

aparcar commented:
If you figure it out feel free to push it

Ansuel committed:
build: make find_md5 reproducible with AUTOREMOVE

While experimenting with the AUTOREMOVE option in search of a way to use
prebuilt host tools in different buildroot, it was discovered that the
md5 generated by find_md5 in depends.mk is not reproducible.

Currently the hash is generated by the path of the file in addition to
the file mod time. Out of confusion, probably, there was an idea that
such command was used on the package build_dir. Reality is that this
command is run on the package files. (Makefile, patches, src)

This is problematic because the package Makefile (for example) change at
each git clone and base the hash on the Makefile mtime doesn't really
reflect if the Makefile actually changes across a buildroot or not.

A better approach is to generate an hash of each file and then generate
an hash on the sort hash list. This way we remove the problem of git
clone setting a wrong mtime while keeping the integrity of checking if a
file changed for the package as any change will result in a different
hash.

Introduce a new kind of find_md5 function, find_md5_reproducible that
apply this new logic and limit it only with AUTOREMOVE option set to
prevent any kind of slowdown due to additional hash generation.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
tools: add option BUILD_ALL_HOST_TOOLS to compile all host tools

Add option to compile all host tools even if not needed.
This can be useful to prepare a universal precompiled host tools
archive to use in another buildroot and speedup compilation.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@aparcar I'm pushing some testing commit also to check the speedup and how it works with a real implementation
(and also if you want to review the changes)

Ansuel commented:
@aparcar Well here is the problem with this... github limitation... 
container can't be pushed by pull request. The github token to make the login works only in read mode for pull request and is denied to push them.
The alternative would be using a PAT (personal access token) and pass it as a secret to test this but I don't want to do that if necessary...

So currently the only way (excluding the PAT way) to test this would be merge the commits and make the github action run a tools worksflow action and (in theory) the container should be correctly pushed as an Organization package.

We can also consider to push the container to docker.io but we need to use secrets and don't know if openwrt have an account on docker.io 

Aside from that with the workaround this works as i tested this on my repo. https://github.com/Ansuel/openwrt/actions/runs/3028914182 As you can see this use ghcr.io/ansuel/tools:latest container and the tools are recycled. 

So yes the only problem here is actually testing the container push for the organization. Don't know how to proceed. Wonder if you have some hint on what to do...
(the documentation about this problem online is very limited... the only source of this limitation is here https://stackoverflow.com/questions/70646920/github-token-permission-denied-write-package-when-build-and-push-docker-in-githu where they say that for pr you have to use PAT and github token can't be used. This makes sense since someone with a pr can abuse this and overwrite organization package and do all sort of mess. So i'm pretty positive that an action triggered by a push will work and actually push the docker image correctly but i'm not 100% sure as no idea how to test... Create a test organization maybe?)

Ansuel commented:
OOOOk had this bright idea of took part to my need handy organization.
I present you testone234567.

https://github.com/testone234567

Anyway I did some test and I can confirm now that it's true.
- github packages (our tools container, that's what they call it) can only be pushed by "push" event. I tested this here https://github.com/testone234567/openwrt/runs/8288241660?check_suite_focus=true
- workflow can use package that are private without using login actions or other kind of thing

Here all the action I tested https://github.com/testone234567/openwrt/actions

With these 2 test I can confirm that this can work so we just need to take a decision if the workaround with the stamp file I'm using is acceptable. Aside from that when this will be pushed everything should work as intended.

I will leave my glorious organization up for a few days till I get some feedback about all of this.

xabolcs commented:
> Well here is the problem with this... github limitation...
> container can't be pushed by pull request. The github token to make the login works only in read mode for pull request and is denied to push them.
> [...]
> As you can see this use `ghcr.io/ansuel/tools:latest` container and the tools are recycled.

Hi @Ansuel,
thats what I wanted to propose too!

Instead of `ghcr.io/openwrt/tools:latest` just use one of the [`github`'s context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context):
* for `master` builds use `tags: ghcr.io/${{ github.repository_owner }}/tools:latest` or even `tags: ghcr.io/${{ github.repository }}/tools:latest` if `github.repository_owner` is not `openwrt`.
* for pull request builds just use `tags: ghcr.io/${{ github.actor }}/tools:latest` or even `tags: ghcr.io/${{ github.actor }}/openwrt/tools:latest` if `github.actor` is not `openwrt`.

Ansuel commented:
@xabolcs The thing is why. The tools workflow should not create a container for pr. That should not be part of the tests for a pr related to tools changes.

As implemented now (and this is actually one of the final version) the container jobs are run only on push. And the container that is created has to be an Organization one as kernel workflow will use that to test kernel build.

Applying the logic of creating containers for each pr/user would result in them no actually used as the kernel workflow will still use the Organization one anyway.

My only problem was with the fact that online and on any documentation it's said that "use the GITHUB_TOKEN and everything will work" but this is obviously wrong as I discovered here and I just tested how to make things work with a testing organization :D

xabolcs commented:
@Ansuel 

I tought you want to test `tools` container push to ghcr on PRs.
If it's not the case and you just want to run some extra gh action after git.openwrt.org -> github sync then just use my `master` suggestion.


In that case it would work for the main repository and also for the forks repositories.

And I talked about the `tags` in the docker buildx step (at the end of the yaml).

Ansuel commented:
Ok I see your point... currently we hardcode this to openwrt and the workflow will only work here...
by using `github.repository_owner` we can permit this to work also on forked repo.

We are not interested in generating containers with pr, as you can see in the yaml they will only run on push events (with tools changes)

xabolcs commented:
> by using `github.repository_owner` we can permit this to work also on forked repo.

Or you can save a few CPU cycles if you filter those Docker steps in the end with `github.repository_owner == 'openwrt'`.

As its doesn' make much sense to build the cobtainer if someone sync his/her fork's master branch.

aparcar commented on include/depends.mk:
```diff
@@ -11,7 +11,7 @@
 DEP_FINDPARAMS := -x "*/.svn*" -x ".*" -x "*:*" -x "*\!*" -x "* *" -x "*\\\#*" -x "*/.*_check" -x "*/.*.swp" -x "*/.pkgdir*"
 
-find_md5=find $(wildcard $(1)) -type f $(patsubst -x,-and -not -path,$(DEP_FINDPARAMS) $(2)) -printf "%p%T@\n" | sort | $(MKHASH) md5
+find_md5=find $(wildcard $(1)) -type f $(patsubst -x,-and -not -path,$(DEP_FINDPARAMS) $(2)) -printf "%p%T@\n" | sort | awk -v srch="$(TOPDIR)" -v repl="" '{ sub(srch,repl,out); print out }' | $(MKHASH) md5
```
```suggestion
find_md5=find $(wildcard $(1)) -type f $(patsubst -x,-and -not -path,$(DEP_FINDPARAMS) $(2)) -printf "%p%T@\n" | sort | sed 's#$(TOPDIR)##' | $(MKHASH) md5
```

Ansuel replied:
I used awk to handle situation with special char in TOPDIR like `/home/Ansuel#/`... 
this will result in 
sed `'s#/home/Ansuel#/##'` and sed will fail

aparcar replied:
What would happen if they have `"` in the path?

Ansuel replied:
Mhh can a directory have `"` in the name? But again guess we need take a decision on this... Probably I'm too paranoic on what shittery an user may have in $TOPDIR

aparcar replied:
In any case, we can also use " as separator for sed

Ansuel replied:
big idea! can we use a whitespace as separator? as we enforce no whitespace in the openwrt path?

Ansuel replied:
@aparcar should we use something like this?
`sed s$'\001'$TOPDIR$'\001'$'\001'`

Ansuel replied:
```
find $(pwd)/tools/flock/ -type f | sort | sed s$'\032'$TOPDIR$'\032'$'\032'
/tools/flock/Makefile
/tools/flock/src/flock.c
```

`\032` should be whitespace

aparcar replied:
@jow- thoughts?

Ansuel replied:
@aparcar wait i notice the problem was another so i'm pushing a real fix... ignore this change since it's REALLY WRONG.

Ansuel replied:
(check the new commit find_md5 if you want to know the findings...)

aparcar commented on .github/workflows/Dockerfile.tools:
```diff
@@ -0,0 +1,3 @@
+FROM registry.gitlab.com/openwrt/buildbot/buildworker-3.4.1
+
+COPY --chown=buildbot:buildbot tools.tar /tools.tar
```
```suggestion
COPY --chown=buildbot:buildbot tools.tar /whereever/
```

Why would you want to have this shipped as a tar and not unpacked?

Ansuel replied:
I decided to take the tar route to be safe on link handling and other timestamps.

On the kernel.yaml workflow we don't link but we extract them.

If you want I can do one more test with the copy and link but currently the extraction and make tools is less than a minute.

aparcar replied:
Right, since we don't store all files but only the tools in there!

aparcar commented on tools/Makefile:
```diff
@@ -26,17 +26,31 @@ tools-y += e2fsprogs fakeroot findutils firmware-utils flex gengetopt
-tools-$(CONFIG_USES_MINOR) += kernel2minor
-tools-$(CONFIG_USE_SPARSE) += sparse
-tools-$(CONFIG_USE_LLVM_BUILD) += llvm-bpf
+ifeq ($(CONFIG_BUILD_ALL_HOST_TOOLS),)
+  tools-$(BUILD_B43_TOOLS) += b43-tools
+  tools-$(BUILD_TOOLCHAIN) += expat gmp mpc mpfr
+  tools-$(BUILD_ISL) += isl
+  tools-$(CONFIG_TARGET_apm821xx)$(CONFIG_TARGET_gemini) += genext2fs
+  tools-$(CONFIG_TARGET_ath79) += lzma-old squashfs
+  tools-$(CONFIG_TARGET_mxs) += elftosb sdimage
+  tools-$(CONFIG_TARGET_realtek) += 7z
+  tools-$(CONFIG_TARGET_tegra) += cbootimage cbootimage-configs
+  tools-$(CONFIG_USES_MINOR) += kernel2minor
+  tools-$(CONFIG_USE_SPARSE) += sparse
+  tools-$(CONFIG_USE_LLVM_BUILD) += llvm-bpf
+else
+  tools-y += b43-tools
+  tools-y += expat gmp mpc mpfr
+  tools-y += isl
+  tools-y += genext2fs
+  tools-y += lzma-old squashfs
+  tools-y += elftosb sdimage
+  tools-y += 7z
+  tools-y += cbootimage cbootimage-configs
+  tools-y += kernel2minor
+  tools-y += sparse
+  tools-y += llvm-bpf
+endif
```
```suggestion
tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS )$(BUILD_B43_TOOLS),y) += b43-tools
tools-$(BUILD_TOOLCHAIN) += expat gmp mpc mpfr
tools-$(BUILD_ISL) += isl
tools-$(CONFIG_TARGET_apm821xx)$(CONFIG_TARGET_gemini) += genext2fs
tools-$(CONFIG_TARGET_ath79) += lzma-old squashfs
tools-$(CONFIG_TARGET_mxs) += elftosb sdimage
tools-$(CONFIG_TARGET_realtek) += 7z
tools-$(CONFIG_TARGET_tegra) += cbootimage cbootimage-configs
tools-$(CONFIG_USES_MINOR) += kernel2minor
tools-$(CONFIG_USE_SPARSE) += sparse
tools-$(CONFIG_USE_LLVM_BUILD) += llvm-bpf
```

aparcar replied:
It's an incomplete example but shouldn't that work, too? This way you don't duplicate it all.

Ansuel replied:
@aparcar ok i can change this to the if logic. Nice idea.

Ansuel replied:
@aparcar can you check if now the logic is correct?

Ansuel commented:
(currently testing the new implementation here https://github.com/testone234567/openwrt/actions/runs/3039956845/jobs/4895472940)

Ansuel commented:
Ok good the new implementation works correctly https://github.com/testone234567/openwrt/actions/runs/3040192171/jobs/4895972922 (tools compilation only 30 seconds)

Ansuel commented:
@aparcar can we merge this?

xabolcs commented:
@Ansuel, @aparcar 

Could this somehow [documented on openwrt.org](https://openwrt.org/docs/guide-developer/start#using_the_toolchain) to speed up local builds?

Ansuel commented:
@xabolcs the container won't be public so what should we document?

xabolcs commented:
> the container won't be public ... 

It would be public. Just pushed this PR to my `master` branch:
> ![image](https://user-images.githubusercontent.com/568751/190999388-89197680-a5a8-4b09-964d-b3f5e05324a9.png)

----

I see the value adding a new "Building system setup Docker" element in the [Toolchain developer guide start](https://openwrt.org/docs/guide-developer/toolchain/start) list:
* just pull the OpenWrt buildbot Docker image and the prerequisites is done without installing any package to host
* just pull another (this `tools`) Docker image and little part of the build work is done
* @mpratt14 's related work of "proposal for a tarball for a prebuilt set of build tools" is also a nice candidate for that Wiki page

Of course this documentation shouldn't be a recommended way of building OpenWrt, but an advanced alternative. :slightly_smiling_face:

Ansuel commented:
Wth did i do?.... sorry let me restore...

xabolcs commented:
> Wth did i do?.... sorry let me restore...

:thinking: 

Nothing wrong!  The screenshot is **from my fork**! The GH action did a nice work: it built and published the `tools` image under my ghcr account.

Ansuel commented:
> 

no sorry was referring on deleting your comment ahahha

Anyway from my test organization package are published private by default... so we should decide about that.

Ansuel commented:
Also can you confirm this works on your pr? or the tools are recompiled?

xabolcs commented:
> Also can you confirm this works on your pr? or the tools are recompiled?

Force pushed my fork, let's see ...

xabolcs commented:
The `Build tools BuildBot Container` step took 18m 54s as my first push ... the tools were recompiled
```
Run make tools/install -j$(nproc) BUILD_LOG=1
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/busybox/Makefile' has a build dependency on 'libpam', which does not exist
WARNING: Makefile 'package/boot/kexec-tools/Makefile' has a dependency on 'liblzma', which does not exist
WARNING: Makefile 'package/network/services/lldpd/Makefile' has a dependency on 'libnetsnmp', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libpam', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libpam', which does not exist
make[2]: Entering directory '/__w/openwrt/openwrt/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/__w/openwrt/openwrt/openwrt/scripts/config'
time: target/linux/prereq#1.27#0.07#1.34
 make[1] tools/install
 make[2] -C tools/flock clean-build
 make[2] -C tools/flock compile
 make[2] -C tools/xz clean-build
 make[2] -C tools/xz compile
<SNIP>
 make[2] -C tools/squashfskit4 clean-build
 make[2] -C tools/squashfskit4 compile
 make[2] -C tools/lzma-old clean-build
 make[2] -C tools/lzma-old compile
 make[2] -C tools/elfutils clean-build
 make[2] -C tools/elfutils compile
 make[2] -C tools/squashfs clean-build
 make[2] -C tools/squashfs compile
```

Ansuel commented:
@xabolcs this is the tools workflow, it's correct that tools are recompiled... you should create a pr that makes changes to the kernel to test this :D

xabolcs commented:
> you should create a pr that makes changes to the kernel to test this :D

Sorry for being noob! :sweat_smile:

Ansuel commented:
Now that the token is restricted by default, I rebased this on top of master and added the required permissions

mpratt14 commented:
even if we don't like the tools tarball, I believe we should have a robust solution in the build system to prevent rebuilds

and by the way, we can add the tarball without hosting it or offering it for download, by removing `default BUILDBOT` from config

Ansuel commented:
@mpratt14 the idea of this is to have a working base and then we can include all the tools improvement and things will just work on top of this.

The idea of shipping a container with the tar in it removes extra step since everything will run in a container anyway :D (one less download)

mpratt14 commented:
yeah sure, of course I like all the ideas here

just the one picky thing, lets fix `find_md5` instead of duplicating it

however that takes a few complex commits to achieve as we discovered, so it can be taken care of later if you want to rush with the main idea

Ansuel commented:
@mpratt14 if you check this the find_md5 is fundamentally wrong... but adding change to it may cause longer compile times so for now it was suggested to add a variant and limit it to CONFIG_AUTOREMOVE that ideally is used only by buildbot/CI and not in normal compilation where time is important.

xabolcs commented:
> Also can you confirm this works on your pr? or the tools are recompiled?

I'm linking here my GH actions, as I don't know what should I see :see_no_evil: :sweat_smile: 
* [first run](https://github.com/xabolcs/openwrt/actions/runs/3082381059)
* [second run](https://github.com/xabolcs/openwrt/actions/runs/3083367189), after your rebase

Ansuel commented:
@xabolcs what you should see are the 20 second to build tools instead of 18-20 minutes :D

mpratt14 commented:
> find_md5 is fundamentally wrong

I agree in general, but it does help in a case where Makefile is changed and then put back to how it was

it also allows for not needing to disable CONFIG_AUTOREBUILD when using prebuilt tools (i want to check that again)

the difference in time is not great, something like 30 sec vs 160 sec

xabolcs commented:
> the 20 second to build tools instead of 18-20 minutes :D

Exactly! Thanks! :grinning: 

In the "Build Kernel with external toolchain" action the "Build tools" step is now ~ 20 s
And the output is:
```
Run make tools/install -j$(nproc) BUILD_LOG=1
make[2]: Entering directory '/__w/openwrt/openwrt/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/__w/openwrt/openwrt/openwrt/scripts/config'
time: target/linux/prereq#1.18#0.08#1.25
 make[1] tools/install
 make[2] -C tools/flock compile
 make[2] -C tools/xz compile
 make[2] -C tools/sed compile
 make[2] -C tools/patch compile
 make[2] -C tools/tar compile
 make[2] -C tools/m4 compile
 make[2] -C tools/autoconf-archive compile
 make[2] -C tools/cpio compile
 make[2] -C tools/ninja compile
 make[2] -C tools/mtools compile
 make[2] -C tools/lzma compile
<SNIP>
 make[2] -C tools/bc compile
 make[2] -C tools/padjffs2 compile
 make[2] -C tools/mkimage compile
 make[2] -C tools/b43-tools clean-build
 make[2] -C tools/quilt compile
 make[2] -C tools/b43-tools compile
```

Compared to [my comment above](/openwrt/openwrt/pull/10421#issuecomment-1250932731): 
> The Build tools BuildBot Container step took 18m 54s as my first push

oliv3r commented:
Just skimmed over this thread, it's super interesting, but I surely hope these will be public containers, like the openwrt package/image builder containers. Doing local development, and having stuff rebuild is just frustrating (though I suppose in my case, it only happens every once in a while, due to of course caching the built binaries). I should be able to 'git clone; docker run <container>; docker run ... make (or docker run /bin/sh ... make) so that all my builds are done using docker, and the tools are practically never built (if not needed, doing a git pull could trigger _some_ tools to be rebuilt as they where updated, but that should be an exception rather then the rule.

Currently, I actually use the openwrt package builder docker container, which already helps with half the per-requisites :)

Ansuel commented:
@oliv3r problem of the container is that it will just have the tools prebuilt tools so you will still have to do some extra step... need to check if that can be improved... something like shipping a prepare script? But it's an hack on top of another hack.

Anyway I hope this can be moved further so we can experiment more with github packages... (making the container public it's really one click)
